internal/provider/data_secret_keys.go

```diff
@@ -0,0 +1,138 @@
+			Description:     "The keys of all the secrets.",
+			DescriptionKind: tfprotov5.StringKindMarkdown,
+			Computed:        true,
+		},
+	}
+
+	return &tfprotov5.Schema{
+		Version: 1,
+		Block: &tfprotov5.SchemaBlock{
+			Description:     "Data source to interpolate any variables or env variables into their final values.",
+			DescriptionKind: tfprotov5.StringKindMarkdown,
+			Attributes:      attrs,
+		},
+	}
+}
+
+func (d *dataSecretKeys) Validate(ctx context.Context, config map[string]tftypes.Value) ([]*tfprotov5.Diagnostic, error) {
+	inputEnvVariables := extractMapFromConfig(config, "input_env_variables")
+	inputSecretKeys := extractSetFromConfig(config, "input_secret_keys")
+
+	errors := make([]*tfprotov5.Diagnostic, 0)
+	for key, _ := range inputEnvVariables {
+		if !validEnvVariableKey(key) {
+			errors = append(errors, &tfprotov5.Diagnostic{
+				Severity: tfprotov5.DiagnosticSeverityError,
+				Summary:  fmt.Sprintf("Invalid environment variable key: %s", key),
+				Detail:   "An environment variable key can only contain letters, numbers, and the underscore character. It also can not begin with a number.",
+			})
+		}
+	}
+	for _, key := range inputSecretKeys {
+		if !validEnvVariableKey(extractStringFromTfValue(key)) {
+			errors = append(errors, &tfprotov5.Diagnostic{
+				Severity: tfprotov5.DiagnosticSeverityError,
+				Summary:  fmt.Sprintf("Invalid environment variable key: %s", key),
+				Detail:   "An environment variable key can only contain letters, numbers, and the underscore character. It also can not begin with a number.",
+			})
+		}
+	}
+
+	return errors, nil
+}
+
+func (d *dataSecretKeys) Read(ctx context.Context, config map[string]tftypes.Value) (map[string]tftypes.Value, []*tfprotov5.Diagnostic, error) {
+	inputEnvVariables := extractMapFromConfig(config, "input_env_variables")
+	inputSecretKeys := extractSetFromConfig(config, "input_secret_keys")
+
+	tflog.Debug(ctx, "input_env_variables", inputEnvVariables)
+	tflog.Debug(ctx, "input_secret_keys", inputSecretKeys)
+
+	// make sure we copy these so our changes below don't affect the original values
+	envVariables := copyMap(inputEnvVariables)
+	secretKeys := copySet(inputSecretKeys)
+
+	regexPattern := "{{\\s?%s\\s?}}"
+
+	// loop through and determine if any of the environment variables contain interpolation using any of the secret keys
+	//   if they do, add their keys to the final set of secret keys
+	for _, secretKey := range inputSecretKeys {
+		regex := regexp.MustCompile(fmt.Sprintf(regexPattern, extractStringFromTfValue(secretKey)))
+		// first try and replace in the env variables
+		for k, v := range envVariables {
+			if found := regex.MatchString(extractStringFromTfValue(v)); found {
+				secretKeys = append(secretKeys, tftypes.NewValue(tftypes.String, k))
+			}
+		}
+	}
+
+	// calculate the unique id for this data source based on a hash of the resulting env variables and secrets
+	id := d.HashFromValues(envVariables, secretKeys)
```

Comment:
Since this data source primarily focuses on secret keys, I would make the id a hash of the secret keys instead of the input values. (Read another way: We don't care if the input changes, only if the output changes)

Reply:
Done